core/trino-main/src/main/java/io/trino/server/NodeStateManager.java

```diff
@@ -69,61 +75,114 @@ public NodeStateManager(
         this.gracePeriod = serverConfig.getGracePeriod();
     }
 
-    public synchronized void requestShutdown()
+    public NodeState getServerState()
     {
-        log.info("Shutdown requested");
+        return nodeState.get();
+    }
 
-        if (isCoordinator) {
-            throw new UnsupportedOperationException("Cannot shutdown coordinator");
-        }
+    /*
+    Below is a diagram with possible states and transitions
+
+    @startuml
+    [*] --> ACTIVE
+    note "state INACTIVE is not used internally\nis only used when the service is unavailable " as a
+    ACTIVE --> SHUTTING_DOWN : shutdown
+    ACTIVE --> DRAINING : drain
+    DRAINING --> ACTIVE: reactivate
+    DRAINING --> DRAINED
+    DRAINING --> SHUTTING_DOWN : gracefulShutdown
+    DRAINED --> ACTIVE: reactivate
+    DRAINED --> SHUTTING_DOWN : terminate
+    SHUTTING_DOWN --> [*]
+    @enduml
 
-        if (shutdownRequested) {
+    NOTE: SHUTTING_DOWN is treated as one-way transition to be 100% backwards compatible.
+    */
+    public synchronized void transitionState(NodeState state)
```

Comment:
Did you think of using `StateMachine` as implementation block here. Like we do in many places. E.g `QueryStateMachine`

Reply:
Regarding the low level method method and exposing specific `transitionTo... ` methods publicly. This will require adding the `switch ...case` in the `io.trino.server.ServerInfoResource#updateState`. Do you think it brings us any value?

Reply:
Regarding the `StateMachine` - I have not seen it before. Interesting idea, but would require me learning how to use it, and rewriting `NodeStateManager`. 

Is it much better or easier to use `StateMachine` instead of hand written code in `NodeStateManager`? Can it be done later as a refactoring task?

The thing is, the code presented here is already in production environments and is well tested and battle-proven. Any changes would require careful analysis and testing.

Reply:
`StateMachine` is kinda trivial. It is just a building block wrapping atomic with a state with nicer interface. It also adds tooling for attaching state change listeners, but you do not need that here I guess